lib/ex_aws/credentials_ini/file.ex

```diff
@@ -4,12 +4,120 @@ if Code.ensure_loaded?(ConfigParser) do
       shared_credentials = profile_from_shared_credentials(profile_name)
       config_credentials = profile_from_config(profile_name)
-      Map.merge(config_credentials, shared_credentials)
+      merge_credentials(config_credentials, shared_credentials)
+    end
+
+    defp merge_credentials(
+           %{
+             sso_start_url: sso_start_url,
+             sso_account_id: sso_account_id,
+             sso_role_name: sso_role_name
+           } = config_credentials,
+           _shared_credentials
+         ) do
+      case get_sso_role_credentials(sso_start_url, sso_account_id, sso_role_name) do
+        {:ok, sso_creds} -> {:ok, Map.merge(sso_creds, config_credentials)}
+        {:error, _} = err -> err
+      end
+    end
+
+    defp merge_credentials(config_credentials, shared_credentials) do
+      {:ok, Map.merge(config_credentials, shared_credentials)}
+    end
+
+    defp get_sso_role_credentials(sso_start_url, sso_account_id, sso_role_name) do
+      with {_, {:ok, content_raw}} <-
+             {:read, File.read(get_sso_cache_file(sso_start_url))},
+           {_, {:ok, %{"expiresAt" => expires_at} = content}} <-
+             {:decode, Jason.decode(content_raw)},
+           {_, :ok} <-
+             {:expiration, check_sso_expiration(expires_at)},
+           {_, {:ok, sso_creds}} <-
+             {:sso_creds, request_sso_role_credentials(content, sso_account_id, sso_role_name)},
+           {_, {:ok, reformatted_creds}} <-
+             {:rename, rename_sso_credential_keys(sso_creds)} do
+        {:ok, reformatted_creds}
+      else
+        {:read, {:error, error}} -> {:error, "Could not read SSO cache file: #{error}"}
+        {:decode, _} -> {:error, "SSO cache file contains invalid json"}
+        {:expiration, error} -> error
+        {:sso_creds, error} -> error
+        {:rename, error} -> error
+      end
+    end
+
+    defp get_sso_cache_file(sso_start_url) do
+      hash = :crypto.hash(:sha, sso_start_url) |> Base.encode16() |> String.downcase()
+
+      System.user_home()
+      |> Path.join(".aws/sso/cache/#{hash}.json")
+    end
+
+    defp check_sso_expiration(expires_at_str) do
+      with {_, {:ok, expires_at, _}} <- {:timestamp, DateTime.from_iso8601(expires_at_str)},
+           {_, :gt} <- {:expires, DateTime.compare(expires_at, DateTime.utc_now())} do
+        :ok
+      else
+        {:timestamp, {:error, err}} ->
+          {:error, "SSO cache file has invalid expiration format: #{err}"}
+
+        {:expires, _} ->
+          {:error, "SSO access token is expired, refresh the token with `aws sso login`"}
+      end
+    end
+
+    defp request_sso_role_credentials(
+           %{"accessToken" => access_token, "region" => region},
+           account_id,
+           role_name
+         ) do
+      with config <- ExAws.Config.http_config(:sso),
+           {:ok, %{status_code: 200, headers: _headers, body: body_raw}} <-
+             config[:http_client].request(
```

Comment:
I think you meant to wrap each side of this in a `{:request, ...}` tuple, right?

As this is currently written, nothing will ever get caught by
```
      else
        {:request, {_, %{status_code: status_code} = resp}} ->
          {:error, "SSO role credentials request responded with #{status_code}: #{resp}"}
```